### PR TITLE
[WIP] Add initial spark on k8s with volcano

### PR DESCRIPTION
**Pull request by Yikun**
Co-authored-by: Klaus Ma <klaus1982.cn@gmail.com>, william-wang <wang.platform@gmail.com>

TODO:
- Rewrite Volcano scheduling java code using scala.
- Add a generic interface for batch scheduler.
- Make the volcano scheduling as default impletation of generic k8s batch scheduler.

### CODE REVIEW

**Comment by Yikun** [comment]
cc @holdenk

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/Config.scala [hunk 1]
```diff
@@ -492,6 +493,8 @@ private[spark] object Config extends Logging {
       .createWithDefault(false)
 
   val KUBERNETES_NODE_SELECTOR_PREFIX = "spark.kubernetes.node.selector."
+  val KUBERNETES_DRIVER_NODE_SELECTOR_PREFIX = "spark.kubernetes.driver.node.selector."
+  val KUBERNETES_EXECUTOR_NODE_SELECTOR_PREFIX = "spark.kubernetes.executor.node.selector."
```
Comment (holdenk) [inline resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/Config.scala:497]: These changes, while I think totally having a selector for exec and driver make sense, we could separate that out into a standalone PR.

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/Config.scala [hunk 2]
```diff
@@ -571,4 +574,35 @@ private[spark] object Config extends Logging {
...
+
+  val KUBERNETES_VOLCANO_PODGROUP_ANNOTATION_PREFIX =
+    "spark.kubernetes.volcano.podgroup.annotation."
+
+  val KUBERNETES_VOLCANO_PODGROUP_MEMORY = ConfigBuilder("spark.kubernetes.volcano.podgroup.memory")
+    .doc("Amount of memory to use for the podgroup minResource, in MiB unless otherwise specified.")
+    .bytesConf(ByteUnit.MiB)
+    .createWithDefaultString("3g")
+
+  val KUBERNETES_VOLCANO_PODGROUP_CPU = ConfigBuilder("spark.kubernetes.volcano.podgroup.cpu")
```
Comment (holdenk) [inline resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/Config.scala:604]: Could we have min as part of this? e.g. `spark.kubernetes.volcano.podgroup.min.cpu` & `spark.kubernetes.volcano.podgroup.min.memory`. That way I think its clearer what this is used for and not confused with the other CPU & memory settings we have.

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/features/BasicDriverFeatureStep.scala [hunk 3]
```diff
@@ -138,6 +138,8 @@ private[spark] class BasicDriverFeatureStep(conf: KubernetesDriverConf)
         .endResources()
       .build()
 
+    val driverSpecifiedLabel = KubernetesUtils.parsePrefixedKeyValuePairs(conf.sparkConf, KUBERNETES_DRIVER_NODE_SELECTOR_PREFIX)
+
```
Comment (holdenk) [inline resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/features/BasicDriverFeatureStep.scala]: I think this is reasonable, but I'd push it out to another PR & JIRA (just minimizing the change for upstream review).

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/features/BasicExecutorFeatureStep.scala [hunk 4]
```diff
@@ -258,6 +258,10 @@ private[spark] class BasicExecutorFeatureStep(
         .withUid(pod.getMetadata.getUid)
         .build()
     }
+
+    val executorSpecifiedLabel = KubernetesUtils.parsePrefixedKeyValuePairs(kubernetesConf.sparkConf,
+      KUBERNETES_EXECUTOR_NODE_SELECTOR_PREFIX)
```
Comment (holdenk) [inline resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/features/BasicExecutorFeatureStep.scala]: Same as with the driver change

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/features/VolcanoFeatureStep.scala [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+      .addToAnnotations(v1beta1.POD_GROUP_ANNOTATION, podGroupName)
+      .endMetadata()
+
+    val k8sPod = k8sPodBuilder.build()
+    SparkPod(k8sPod, pod.container)
+  }
+
+  override def getAdditionalPodSystemProperties(): Map[String, String] = Map.empty
+
+  private def getPodResources(): Map[String, Quantity] = {
```
Comment (holdenk) [inline resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/features/VolcanoFeatureStep.scala:60]: Maybe getPodGroupResources unless I've missunderstood.

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/features/VolcanoFeatureStep.scala [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+import sh.volcano.scheduling.{PodGroup, PodGroupSpec, PodGroupStatus, v1beta1}
+import org.apache.spark.deploy.k8s.{KubernetesConf, KubernetesDriverSpecificConf, KubernetesRoleSpecificConf, KubernetesUtils, SparkPod}
+import org.apache.spark.deploy.k8s.Config._
+
+private[spark] class VolcanoFeatureStep(
+                                         kubernetesConf: KubernetesConf[_ <: KubernetesRoleSpecificConf])
+  extends KubernetesFeatureConfigStep {
+
+  private val conf = kubernetesConf.sparkConf
+  private val podGroupName = s"${kubernetesConf.appResourceNamePrefix}-podgroup"
```
Comment (holdenk) [inline resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/features/VolcanoFeatureStep.scala]: Similar comment around I think it's find for v1 but for v2 we probably need to do something with resource profiles.

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/features/VolcanoFeatureStep.scala [hunk 7]
```diff
@@ -0,0 +1,102 @@
...
+      conf, KUBERNETES_VOLCANO_PODGROUP_ANNOTATION_PREFIX)
+    val queue = conf.get(KUBERNETES_VOLCANO_QUEUE)
+
+    val meta = new ObjectMeta()
+    meta.setAnnotations(podGroupAnnotations.asJava)
+    meta.setName(podGroupName)
+    meta.setNamespace(kubernetesConf.namespace())
+
+    val podGroupSpec = new PodGroupSpec()
+    podGroupSpec.setMinMember(1)
```
Comment (holdenk) [inline resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/features/VolcanoFeatureStep.scala:85]: Would it make sense to have the minimum number of resources be configurable instead of 1?

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/features/VolcanoFeatureStep.scala [hunk 8]
```diff
@@ -0,0 +1,102 @@
...
+    podGroupSpec.setQueue(queue)
+    podGroupSpec.setPriorityClassName("")
+    podGroupSpec.setMinResources(getPodResources().asJava)
+
+    val podGroupStatus = new PodGroupStatus()
+    podGroupStatus.setPhase("Pending")
```
Comment (holdenk) [inline resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/features/VolcanoFeatureStep.scala:91]: This seems strange, I'm used to the phase being set by the CRD

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/features/VolcanoFeatureStep.scala [hunk 9]
```diff
@@ -0,0 +1,102 @@
...
+
+    val podGroup = new PodGroup()
+    podGroup.setMetadata(meta)
+    podGroup.setSpec(podGroupSpec)
+    podGroup.setStatus(podGroupStatus)
```
Comment (holdenk) [inline resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/features/VolcanoFeatureStep.scala:96]: Same as above, I'm a bit fuzzy on this.

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/scheduler/cluster/k8s/KubernetesExecutorBuilder.scala [hunk 10]
```diff
@@ -46,13 +46,17 @@ private[spark] class KubernetesExecutorBuilder {
...
+    } else Nil
+
     val features = Seq(
       new BasicExecutorFeatureStep(conf, secMgr, resourceProfile),
       new ExecutorKubernetesCredentialsFeatureStep(conf),
       new MountSecretsFeatureStep(conf),
       new EnvSecretsFeatureStep(conf),
       new MountVolumesFeatureStep(conf),
-      new LocalDirsFeatureStep(conf)) ++ userFeatures
+      new LocalDirsFeatureStep(conf)) ++ userFeatures ++ volcanoFeature
```
Comment (holdenk) [inline resource-managers/kubernetes/core/src/main/scala/org/apache/spark/scheduler/cluster/k8s/KubernetesExecutorBuilder.scala]: I would put volacnoFeature before userFeatures since we want the user to be able to do pretty much anything as an escape hatch.

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/features/VolcanoFeatureStep.scala [hunk 11]
```diff
@@ -0,0 +1,102 @@
...
+private[spark] class VolcanoFeatureStep(
+                                         kubernetesConf: KubernetesConf[_ <: KubernetesRoleSpecificConf])
+  extends KubernetesFeatureConfigStep {
+
+  private val conf = kubernetesConf.sparkConf
+  private val podGroupName = s"${kubernetesConf.appResourceNamePrefix}-podgroup"
+  private val deployModeKey = "spark.submit.deployMode"
+  private val clusterMode = "cluster"
+
+  override def configurePod(pod: SparkPod): SparkPod = {
```
Comment (holdenk) [inline resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/features/VolcanoFeatureStep.scala:36]: This looks reasonable to me, just more as a note to myself for something I want to explore here is using a replicaset per resource profile.

**Review by holdenk — COMMENTED** [review]
Thank you so much for making this. I've left some initial questions to try and better understand the code.

Also (and I think this is OK for the initial one), but does it make sense to have one podgroup per resource profile or one general podgroup? I go back and forth on this, just thinking out loud but for example a resource profile that requires X GPUs before launch might be something we want users to be able to express (e.g. if they were using this to power a TF job with feed_dict).

**Comment by holdenk** [comment]
I'm going to suggest merging this PR and then we can collaborate in the branch directly.

**Comment by holdenk** [comment]
I tried to build this and I ran into something which suggested maybe there might a file that isn't tracked by git yet, can you check your git status?

```
[INFO] compiling 44 Scala sources and 7 Java sources to /home/holden/repos/spark/resource-managers/kubernetes/core/target/scala-2.12/classes ...
[ERROR] /home/holden/repos/spark/resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/features/VolcanoFeatureStep.scala:26: object KubernetesDriverSpecificConf is not a member of package org.apache.spark.deploy.k8s
[ERROR] /home/holden/repos/spark/resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/features/VolcanoFeatureStep.scala:30: not found: type KubernetesRoleSpecificConf
[ERROR] /home/holden/repos/spark/resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/features/VolcanoFeatureStep.scala:30: org.apache.spark.deploy.k8s.KubernetesConf does not take type parameters
[ERROR] /home/holden/repos/spark/resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/features/VolcanoFeatureStep.scala:39: not found: type KubernetesDriverSpecificConf
[ERROR] /home/holden/repos/spark/resource-managers/kubernetes/core/src/main/scala/sh/volcano/scheduling/v1beta1.java:44: wrong number of type arguments for io.fabric8.kubernetes.client.dsl.NonNamespaceOperation, should be 3
[ERROR] /home/holden/repos/spark/resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/submit/KubernetesDriverBuilder.scala:46: not found: value volcanoStep
[ERROR] /home/holden/repos/spark/resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/submit/KubernetesDriverBuilder.scala:46: not found: value kubernetesConf
[ERROR] /home/holden/repos/spark/resource-managers/kubernetes/core/src/main/scala/org/apache/spark/scheduler/cluster/k8s/KubernetesExecutorBuilder.scala:50: not found: value volcanoStep
[ERROR] 8 errors found
[INFO] ------------------------------------------------------------------------
[INFO] Reactor Summary for Spark Project Parent POM 3.2.0-SNAPSHOT:

```

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/Config.scala (hunk 1 above)
Comment (Yikun) [reply resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/Config.scala:497]: OK, I submit it and plus some ut in: https://github.com/apache/spark/pull/33283 .

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/features/BasicDriverFeatureStep.scala (hunk 3 above)
Comment (Yikun) [reply resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/features/BasicDriverFeatureStep.scala]: https://github.com/apache/spark/pull/33283

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/features/BasicExecutorFeatureStep.scala (hunk 4 above)
Comment (Yikun) [reply resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/features/BasicExecutorFeatureStep.scala]: https://github.com/apache/spark/pull/33283

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/scheduler/cluster/k8s/KubernetesExecutorBuilder.scala (hunk 10 above)
Comment (Yikun) [reply resource-managers/kubernetes/core/src/main/scala/org/apache/spark/scheduler/cluster/k8s/KubernetesExecutorBuilder.scala]: Yep, it make sense. will address

**Comment by Yikun** [comment]
> I tried to build this and I ran into something which suggested maybe there might a file that isn't tracked by git yet, can you check your git status?

Yes, the orignal code base is 2.4 with io.fabric8.kubernetes-client, it has been addressed in https://github.com/spark-volcano/spark/pull/2/commits/5a7de42aa3529394d146a407549376eaa7ed18ea
